activerecord/lib/active_record/signed_id.rb

```diff
@@ -70,34 +76,12 @@ def find_signed(signed_id, purpose: nil, on_rotation: nil)
       #   User.first.destroy
       #   User.find_signed! signed_id # => ActiveRecord::RecordNotFound
       def find_signed!(signed_id, purpose: nil, on_rotation: nil)
-        if id = signed_id_verifier.verify(signed_id, purpose: combine_signed_id_purposes(purpose), on_rotation: on_rotation)
+        options = { on_rotation: on_rotation }.compact
+        if id = signed_id_verifier.verify(signed_id, purpose: combine_signed_id_purposes(purpose), **options)
           find(id)
         end
       end
 
-      # The verifier instance that all signed ids are generated and verified from. By default, it'll be initialized
-      # with the class-level +signed_id_verifier_secret+, which within \Rails comes from the
-      # Rails.application.key_generator. By default, it's SHA256 for the digest and JSON for the serialization.
-      def signed_id_verifier
-        @signed_id_verifier ||= begin
-          secret = signed_id_verifier_secret
-          secret = secret.call if secret.respond_to?(:call)
-
-          if secret.nil?
-            raise ArgumentError, "You must set ActiveRecord::Base.signed_id_verifier_secret to use signed ids"
-          else
-            ActiveSupport::MessageVerifier.new secret, digest: "SHA256", serializer: JSON, url_safe: true
-          end
-        end
-      end
```

Comment:
`signed_id_verifier_secret` is also a `class_attribute`, which is unfortunate.  That means existing apps could have, for example, set `Post.signed_id_verifier_secret` and `Comment.signed_id_verifier_secret` to different values, and `Post.signed_id_verifier` and `Comment.signed_id_verifier` each have to use those values, respectively.

I have not fully sketched this out, but I would try giving `ActiveRecord::Base` access to `Rails.application.message_verifiers`.  In other words, set something like `ActiveRecord::Base.message_verifiers` in an initializer, and use that instead of `MessageVerifier.new` in this method.

Reply:
@jonathanhefner Good point. I intended to maintain the current configuration; however, doing so would prevent us from migrating to the default configuration.
I tried it here [https://github.com/rails/rails/pull/54474](https://github.com/rails/rails/pull/54474), please take a look.

Reply:
Is there a particular reason for a separate PR?  If not, I think you should add those changes to this PR so that we can consolidate the discussion.

Reply:
I initially thought we might comeback to this solution if the new one doesn't prove to be better. In any case, I've incorporated the changes into this PR.